cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,335 @@
+{
+    USceneComponent::TickComponent(DeltaTime, TickType, ThisTickFunction);
+
+    if (!ticked_once_) {
+        SP_LOG_CURRENT_FUNCTION();
+        initializeDeferred();
+        ticked_once_ = true;
+    }
+
+    // Update this component's pose to match root link component
+    SP_ASSERT(root_link_component_);
+    bool sweep = false;
+    FHitResult* hit_result = nullptr;
+    SetWorldLocationAndRotation(root_link_component_->GetComponentLocation(), root_link_component_->GetComponentRotation(), sweep, hit_result, ETeleportType::None);
+}
+
+std::map<std::string, ArrayDesc> UUrdfRobotComponent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+
+    if (Std::contains(action_components_, "control_joints")) {
+        for (auto joint_component : joint_components_) {
+            action_space.merge(joint_component.second->getActionSpace());
+        }
+    }
+
+    // remove empty map entry
+    action_space.erase("");
+
+    return action_space;
+}
+
+std::map<std::string, ArrayDesc> UUrdfRobotComponent::getObservationSpace() const
+{
+    std::map<std::string, ArrayDesc> observation_space;
+
+    if (Std::contains(observation_components_, "links_location")) {
+        for (auto& link_component : link_components_) {
+            ArrayDesc array_desc;
+            array_desc.low_ = std::numeric_limits<double>::lowest();
+            array_desc.high_ = std::numeric_limits<double>::max();
+            array_desc.shape_ = {3}; // x, y, z in [cm] of each link relative to it's parent
+            array_desc.datatype_ = DataType::Float64;
+            observation_space[link_component.first] = std::move(array_desc);
+        }
+    }
+
+    if (Std::contains(observation_components_, "links_rotation")) {
+        for (auto& link_component : link_components_) {
+            ArrayDesc array_desc;
+            array_desc.low_ = std::numeric_limits<double>::lowest();
+            array_desc.high_ = std::numeric_limits<double>::max();
+            array_desc.shape_ = {3}; // pitch, yaw, roll in [deg] of each link relative to it's parent
+            array_desc.datatype_ = DataType::Float64;
+            observation_space[link_component.first] = std::move(array_desc);
+        }
+    }
+
+    return observation_space;
+}
+
+void UUrdfRobotComponent::applyAction(const std::map<std::string, std::vector<uint8_t>>& actions)
+{
+    std::map <std::string, std::vector<double>> joint_actions;
+
+    for (auto& action : actions) {
+        std::vector<double> action_data = Std::reinterpretAs<double>(action.second);
+        joint_actions[action.first] = action_data;
+    }
+
+    bool assert_if_action_is_inconsistent_with_joint = true;
+    bool assert_if_joint_not_found = true;
+    applyAction(joint_actions, assert_if_joint_not_found, assert_if_action_is_inconsistent_with_joint);
+}
+
+std::map<std::string, std::vector<uint8_t>> UUrdfRobotComponent::getObservation() const
```

Comment:
```
    std::map<std::string, std::vector<uint8_t>> observation;

    if (Std::contains(observation_components_, "link_configurations")) {
        for (auto link_component : link_components_) {
            observation.merge(link_component->getObservation());
        }
    }

    if (Std::contains(observation_components_, "joint_configurations")) {
        for (auto joint_component : joint_components_) {
            observation.merge(joint_component->getObservationSpace());
        }
    }

    return observation_space;
```

Reply:
Even cleaner would be to iterate over `JointComponents` and `LinkComponents` array instead of the maps.

Reply:
Agreed. Less cluttered notation because then we wouldn't need to use `first` and `second` all over the place.